Dev tunnels in Visual Studio for ASP.NET Core projects (#27960)
dotnet/AspNetCore.Docs: aspnetcore/test/dev-tunnels.md

Choose not to use a tunnel by selecting **None** in the flyout. When Visual Studio is restarted, it defaults back to **None**.

## Use a tunnel

When a tunnel is active and Visual Studio runs a web app, the web browser opens to a tunnel URL instead of a localhost URL. The tunnel URL looks like the following example:

`https://0pbvlk3m-7032.usw2.devtunnels.ms`

Now any authenticated user can open the same URL on any other internet-connected device. As long as the project continues to run locally, any device with an internet connection can access the web application that is running on a development machine.

For web projects that have browser support, a warning page is shown on the first request sent to the tunnel URL from each device:

:::image type="content" source="~/test/dev-tunnels/_static/dev-tunnels-notification.png" alt-text="Dev tunnels notification page.":::

After **Continue** is selected, the request is routed to the local web app. This notification page isn't shown for API requests using dev tunnels.

### Use a tunnel to test on a phone or tablet

> To test a web app on an external device like a phone or tablet, navigate to the tunnel URL on the device. To make it easier to reproduce the URL on the external device:

Author: ```suggestion
To test a web app from an external device like a phone or tablet, navigate to the tunnel URL. To make it easier to reproduce the URL on the external device:
```

info: Microsoft.Hosting.Lifetime[14]
      Now listening on: https://localhost:7175
info: Microsoft.Hosting.Lifetime[14]
      Now listening on: http://localhost:5228
info: Microsoft.Hosting.Lifetime[0]
      Application started. Press Ctrl+C to shut down.
info: Microsoft.Hosting.Lifetime[0]
      Hosting environment: Development
info: Microsoft.Hosting.Lifetime[0]
      Content root path: C:\DevTunnelsDemo\MyWebApp
```

For information about how to set up multiple startup projects, see [How to: Set multiple startup projects](/visualstudio/ide/how-to-set-multiple-startup-projects).

## Persistent versus temporary tunnels

A persistent tunnel is one that uses the same URL after exiting and restarting Visual Studio. Having a URL that doesn't change can be useful when integrating a web app with an external service. For example, implementing a GitHub webhook, or developing an API to integrate with a Power Platform app. In such cases, you might need to specify the callback URL to the external service. With a persistent tunnel, the external service URL only needs to be configured once. Using a temporary tunnel, the tunnel URL must be configured each time Visual Studio restarts.

> *Persistent* doesn't mean the tunnel works when Visual Studio isn't open. A tunnel URL connects to the local machine only if:
> 
> * Visual Studio is open.
> * A tunnel is selected as active.
> * The ASP.NET Core project that the tunnel URL connects to is running.

Author: ```suggestion
*Persistent* doesn't mean the tunnel works when Visual Studio isn't open. A tunnel URL connects to the local machine only if the ASP.NET Core project that the tunnel URL connects to is running in Visual Studio.
```

---
title: Dev tunnels in Visual Studio 2022
author: tdykstra
description: Learn how to use dev tunnels in Visual Studio with ASPNET Core apps.
monikerRange: '>= aspnetcore-7.0'
ms.author: tdykstra
ms.date: 12/28/2022
uid: test/tunnels
---
# How to use dev tunnels in Visual Studio 2022 with ASP.NET Core apps

> The *Dev tunnels* feature of Visual Studio 2022 enables ad-hoc connections between machines that can't directly connect to each other. A URL is created that enables any device with an internet connection to connect to an ASP.NET Core project while it runs on localhost.

Reviewer: Can we change "Dev tunnels" to "dev tunnels", the branding for this service is to have it lower case unless it's required to be upper case (for example, tool window names in VS use caps).

Choose not to use a tunnel by selecting **None** in the flyout. When Visual Studio is restarted, it defaults back to **None**.

## Use a tunnel

When a tunnel is active and Visual Studio runs a web app, the web browser opens to a tunnel URL instead of a localhost URL. The tunnel URL looks like the following example:

`https://0pbvlk3m-7032.usw2.devtunnels.ms`

Now any authenticated user can open the same URL on any other internet-connected device. As long as the project continues to run locally, any device with an internet connection can access the web application that is running on a development machine.

For web projects that have browser support, a warning page is shown on the first request sent to the tunnel URL from each device:

:::image type="content" source="~/test/dev-tunnels/_static/dev-tunnels-notification.png" alt-text="Dev Tunnels notification page.":::

After **Continue** is selected, the request is routed to the local web app. This notification page isn't shown for API requests using dev tunnels.

### Use a tunnel to test on a phone or tablet

> To test a web app on an external device like a phone or tablet, navigate to the tunnel URL on the device. To make it easier to reproduce the URL on the external device:

Author: To test on an external device, you do go to the tunnel URL on the device.

> To test a web app on an external device like a phone or tablet, navigate to the tunnel URL on the device. To make it easier to reproduce the URL on the external device:

Reviewer 2: > To test on an external device, you do go to the tunnel URL on the device.

That's ambiguous. What's the tunnel URL on the device? On the device, navigate to the tunnel URL.

What's "tunnel URL on the device" ?

It's much more clear if you drop on the device.

To test a web app on an external device like a phone or tablet, navigate to the tunnel URL.~ on the device.~ To make it easier to reproduce the URL on the external device:

> To test a web app on an external device like a phone or tablet, navigate to the tunnel URL on the device. To make it easier to reproduce the URL on the external device:

Author: ```suggestion
To test a web app from an external device like a phone or tablet, navigate to the tunnel URL. To make it easier to reproduce the URL on the external device:
```

    :::image type="content" source="~/test/dev-tunnels/_static/generate-qr-code.png" alt-text="QR code with button to create it highlighted.":::
* Scan this QR code with a phone or tablet to navigate to the URL.

## Dev Tunnels output window

The **Dev Tunnels** output window shows the port and URL of a tunnel when a project starts using it. Select the **Dev Tunnels** output window by using the **Show output from:** dropdown in the **Output** window.

:::image type="content" source="~/test/dev-tunnels/_static/dev-tunnels-output-window.png" alt-text="Dev Tunnels output window.":::

This window is especially useful for projects that don't open a browser by default. For example, when working with an Azure Function, this may be the easiest way to discover the public URL that is being used by the dev tunnel.

## Dev Tunnels tool window

View and manage dev tunnels in the **Dev Tunnels** tool window:

:::image type="content" source="~/test/dev-tunnels/_static/dev-tunnels-tool-window.png" alt-text="Dev Tunnels tool window.":::

To open this window, Use the **Show Dev Tunnels Window** menu option in the debug dropdown. Or select **View** > **Other Windows** > **Dev Tunnels**.

> From this tool window, create a new tunnel by selecting the green `+` button. When you create a tunnel from the tool window, it isn't automatically set as the active tunnel like it is when a tunnel is created from the debug dropdown.

Reviewer 2: ```suggestion
From the **Dev Tunnels**  window, create a new tunnel by selecting the green `+` button. When a tunnel is created from the **Dev Tunnels** window, it isn't automatically set as the active tunnel like it is when a tunnel is created from the debug dropdown.
```
Then how to you make it the active tunnel?

Reviewer 2: See next comment

Author: Currently you make it active by using the Debug dropdown, later you'll be able to do it in this tool window also.

> Delete a tunnel by using the red `x` button on the right side.

Reviewer 2: ```suggestion
Delete the previous tunnel by using the red `x` button on the right side. When a new tunnel is created, it isn't automatically set as the active tunnel like it is when a tunnel is created from the debug dropdown.
```

## Tunnel URL environment variable support

> The dev tunnels feature provides a way to get the tunnel URL of a project programmatically at run time. When you launch an app that uses a tunnel, Visual Studio creates an environment variable named `VS_TUNNEL_URL`. This is the URL for the tunnel that is used for the current project. It can be useful when you're integrating the application with an external service, and you need the URL to pass to the external service.

Reviewer 2: ```suggestion
The dev tunnels feature provides a way to get the tunnel URL of a project programmatically at run time. When an app is launched that uses a tunnel, Visual Studio creates the environment variable `VS_TUNNEL_URL`. The `VS_TUNNEL_URL` value is the URL for the tunnel that is used for the current project. `VS_TUNNEL_URL` can be useful when integrating the app with an external service, where the tunnel url needs to be passed to the external service.
```

> If multiple ASP.NET Core projects are configured to start in Visual Studio, the app that is starting up will get an environment variable for any project that started before it. The pattern for this variable name is `VS_TUNNEL_URL_{ProjectName}`, where `{ProjectName}` is the name of the other project. For example, consider this example showing two projects set to start:

Reviewer 2: ```suggestion
If multiple ASP.NET Core projects are configured to start in Visual Studio, the app that is starting up gets an environment variable for any project that started before it. The pattern for this variable name is `VS_TUNNEL_URL_{ProjectName}`, where `{ProjectName}` is the name of the other project. For example, consider this example showing two projects set to start:
```

:::image type="content" source="~/test/dev-tunnels/_static/multi-project-start.png" alt-text="Startup projects selection page showing MyWebApi and MyWebApp both starting, in that order.":::

> Since MyWebApi is above MyWebApp, it's started before the MyWebApp project. When the MyWebApi project is started, it will receive its tunnel URL in `VS_TUNNEL_URL`. When the MyWebApp project is started, it will receive its own tunnel URL in `VS_TUNNEL_URL` and the other project's tunnel URL in `VS_TUNNEL_URL_MyWebApi`. 

Reviewer 2: ```suggestion
Since MyWebApi is above MyWebApp, it's started before the MyWebApp project. When the MyWebApi project is started, it receives its tunnel URL in the `VS_TUNNEL_URL` environment variable. When the MyWebApp project is started, it receives its own tunnel URL in `VS_TUNNEL_URL` and the other project's tunnel URL is created in the `VS_TUNNEL_URL_MyWebApi` environment variable. 
```

To illustrate, the following highlighted lines of code have been added to the *Program.cs* file in MyWebApp:

:::code language="csharp" source="~/test/dev-tunnels/sample/DevTunnelsDemo/MyWebApp/Program.cs" id="snippet1" highlight="5-8":::

When the web app is started the console output looks like the following example:

```console
Tunnel URL: https://lxm0ltdt-7175.usw2.devtunnels.ms/
API project tunnel URL: https://lxm0ltdt-7042.usw2.devtunnels.ms/
info: Microsoft.Hosting.Lifetime[14]
      Now listening on: https://localhost:7175
info: Microsoft.Hosting.Lifetime[14]
      Now listening on: http://localhost:5228
info: Microsoft.Hosting.Lifetime[0]
      Application started. Press Ctrl+C to shut down.
info: Microsoft.Hosting.Lifetime[0]
      Hosting environment: Development
info: Microsoft.Hosting.Lifetime[0]
      Content root path: C:\DevTunnelsDemo\MyWebApp
```

For information about how to set up multiple startup projects, see [How to: Set multiple startup projects](/visualstudio/ide/how-to-set-multiple-startup-projects).

## Persistent versus temporary tunnels

> A persistent tunnel is one that uses the same URL after you exit and restart Visual Studio. Having a URL that doesn't change might be useful when you're integrating a web app with an external service. For example, you're implementing a GitHub webhook, or developing an API to integrate with a Power Platform app. In such cases, you need to specify the callback URL to the external service. With a persistent tunnel, you only need to configure the external service with that URL once. With a temporary tunnel, you'd have to reconfigure the external service to point to the new URL each time Visual Studio restarts.

Reviewer 2: ```suggestion
A persistent tunnel is one that uses the same URL after exiting and restarting Visual Studio. Having a URL that doesn't change can be useful when integrating a web app with an external service. For example, implementing a GitHub webhook, or developing an API to integrate with a Power Platform app. In such cases, specify the callback URL to the external service. With a persistent tunnel, the external service URL only needs to be configured once. Using a temporary tunnel, the tunnel URL must be configured each time Visual Studio restarts.
```

> *Persistent* doesn't mean the tunnel works when Visual Studio isn't open. Any tunnel URL will connect to the local machine only if:

Reviewer 2: ```suggestion
*Persistent* doesn't mean the tunnel works when Visual Studio isn't open. A tunnel URL connects to the local machine only if:
```

> * Visual Studio is open.

Reviewer 2: ```suggestion
* Visual Studio is open and the project is running.
```

Author: That's in the third bullet.

Reviewer 2: You don't need it. `The ASP.NET Core project that the tunnel URL connects to is running.` implies that VS is open. I recommend dropping that line or combining it.

* A tunnel is selected as active.
* The ASP.NET Core project that the tunnel URL connects to is running.

> Even if the tunnel is persistent, the tunnel URL doesn't work if the associated project isn't running in Visual Studio.

Reviewer 2: Recommend deleting but if you want to keep
```suggestion
Persistent tunnel URLs don't work if the associated project isn't running in Visual Studio.
```

> 
> Even if the tunnel is persistent, the tunnel URL doesn't work if the associated project isn't running in Visual Studio.

Author: ```suggestion
```

> A temporary tunnel is fine when you only need the dev tunnel URL to work for a short time. For example, you want to share in-progress work on a web app with others, or you want to test an app on an external device. In some cases, it might be best to get a new URL each time Visual Studio starts.

Reviewer 2: ```suggestion
A temporary tunnel is fine when the dev tunnel URL needs to work for a short time. For example, sharing in-progress work on a web app with others, or testing an app on an external device. In some cases, it might be best to get a new URL each time Visual Studio starts.
```
